feat(typeahead): enable/disable options to make selection on tab-event (#3842)

SidekickJohn opened the pull request:
Adds an optional Input to typeahead directive. Initially true. If set to false, tab will just do its default behaviour.

 Closes #3842

# PR Checklist
Before creating new PR, please take a look at checklist below to make sure that you've done everything that needs to be done before we can merge it.

 - [X] read and followed the [CONTRIBUTING.md](https://github.com/valor-software/ngx-bootstrap/blob/development/CONTRIBUTING.md) guide.
 - [X] built and tested the changes locally.
 - [ ] added/updated tests.
 - [X] added/updated API documentation.
 - [X] added/updated demos.

SidekickJohn committed:
feat(typeahead): enable/disable options to make selection on tab-event (#3842)

Adds an optional Input to typeahead directive. Initially true. If set to false, tab will just do its default behaviour.

 Closes #3842

valorkin commented on src/typeahead/typeahead.directive.ts:
```diff
@@ -73,8 +73,11 @@ export class TypeaheadDirective implements OnInit, OnDestroy {
   @Input() typeaheadPhraseDelimiters = '\'"';
+  /** used to enable/disable the selection on TAB Event.   
+   */
+  @Input() typeaheadSelectOnTab: boolean = true;
```
Hey, thanks for PR, all looks good.
But one change is must for merge,
in terms of app consistency and perf, it would be better to have `selectOnTab` in a config
and not in `@Input`

SidekickJohn replied:
Yes you are right. But same goes for "typeaheadSingleWords" and other boolean values... 
Still: I'm trying to change the `selectOnTab` to be a config and will update my PR.
@valorkin is there an example (for the config part) that I could learn from? Or does the Typeahead-directive already have a config somewhere?

SidekickJohn replied:
@valorkin @IlyaSurmay I still need a little advice on how to solve this one as a config. Is there an example on how to put things like this in a config?

valorkin commented:
As a heads up, I remember about this PR

SidekickJohn committed:
Merge remote-tracking branch 'upstream/development' into development
For keeping my fork up to date with remote

SidekickJohn committed:
feat(typeahead): enable/disable options to make selection on tab-event (#3842)

selectOnTab is now moved to a config
Still possible to set config via Input() this is just like in the tabset.component.ts

valorkin commented on src/typeahead/typeahead.directive.ts:
```diff
@@ -77,7 +78,13 @@ export class TypeaheadDirective implements OnInit, OnDestroy {
   /** used to enable/disable the selection on TAB Event.   
    */
-  @Input() typeaheadSelectOnTab: boolean = true;
+  @Input() 
```
Don't rename the field, just set it with assigment
Field rename will break ppl's code

valorkin commented on src/typeahead/typeahead.directive.ts:
```diff
@@ -73,8 +74,17 @@ export class TypeaheadDirective implements OnInit, OnDestroy {
+  /** used to enable/disable the selection on TAB Event.   
+   */
+  @Input() 
+  get selectOnTab(): boolean {
```
I believe this option is fine to be only in config, no need for additional input

SidekickJohn replied:
@valorkin: How can a User of the typeahead-directive chose whether the option is selectOnTab = true or false? When putting the directive on an input-element, how can i decide this when there is no ` @Input()` for it? <br/>
Should I do it like the Datepicker Component and implement a setConfig() Method that reassigns the _config? Then there would still need to be an `@Input()` for the config...

SidekickJohn replied:
@valorkin @Domainv I still don't know what to do right here. For us it is required that the user can decide whether the selectOnTab is true or false in the HTML for each typeahead instance seperately. Could you please be a little more precise what you expect here? Or tell me an example file in an other component where this is solved the way you imagine it.

valorkin commented on src/typeahead/typeahead.directive.ts:
```diff
@@ -140,12 +151,14 @@ export class TypeaheadDirective implements OnInit, OnDestroy {
       viewContainerRef,
       renderer
     );
+    Object.assign(this, config);
```
set it to `this._config`

valorkin reviewed (changes requested):
This great, almost there!
Couple small changes left.

Each @Input adds performance cost, I believe there are no need for input for this option

ludmilanesvitiy commented:
This feature is already implemented and presented on the demo page:
https://valor-software.com/ngx-bootstrap/#/typeahead#selected-first-item
